kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h

```diff
@@ -150,214 +161,151 @@ volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx2(float* target, lv_32fc_t* s
+                                                    unsigned int num_points) {
+  const int eightsPoints = num_points / 8;
+  const int remainder = num_points - 8 * eightsPoints;
+  
+  __m256 xmm_points0, xmm_points1, xmm_result;
+
+  // load complex value into all parts of the register.
+  const __m256 xmm_symbol = _mm256_castpd_ps(_mm256_broadcast_sd((const double*)src0));
+  
+  // Load scalar into all 8 parts of the register
+  const __m256 xmm_scalar = _mm256_broadcast_ss(&scalar);
+  
+  for(int i = 0; i < eightsPoints; ++i){
+    xmm_points0 = _mm256_load_ps((float*)points);
+    xmm_points1 = _mm256_load_ps((float*)(points + 4));
+    points += 8;
+    
+    xmm_result = _mm256_scaled_norm_dist_ps(xmm_symbol, xmm_symbol, xmm_points0, 
+                                            xmm_points1, xmm_scalar);
+    
+    _mm256_store_ps(target, xmm_result);
+    target += 8;
   }
+  
+  const lv_32fc_t symbol = *src0;
+  calculate_scaled_distances(target, symbol, points, scalar, remainder);
 }
 
 #endif /* LV_HAVE_AVX */
 
 
 #ifdef LV_HAVE_SSE3
-#include<xmmintrin.h>
 #include<pmmintrin.h>
+#include<volk/volk_sse3_intrinsics.h>
 
 static inline void
-volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_sse3(float* target, lv_32fc_t* src0, lv_32fc_t* points,
-                                                     float scalar, unsigned int num_points)
+volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_sse3(float* target, lv_32fc_t* src0, 
+                                                     lv_32fc_t* points, float scalar, 
+                                                     unsigned int num_points)
 {
-  const unsigned int num_bytes = num_points*8;
-
-  __m128 xmm1, xmm2, xmm3, xmm4, xmm5, xmm6, xmm7, xmm8;
-
-  lv_32fc_t diff;
-  memset(&diff, 0x0, 2*sizeof(float));
-
-  float sq_dist = 0.0;
-  int bound = num_bytes >> 5;
-  int leftovers0 = (num_bytes >> 4) & 1;
-  int leftovers1 = (num_bytes >> 3) & 1;
-  int i = 0;
-
-  xmm1 = _mm_setzero_ps();
-  xmm1 = _mm_loadl_pi(xmm1, (__m64*)src0);
-  xmm2 = _mm_load_ps((float*)&points[0]);
-  xmm8 = _mm_load1_ps(&scalar);
-  xmm1 = _mm_movelh_ps(xmm1, xmm1);
-  xmm3 = _mm_load_ps((float*)&points[2]);
+  __m128 xmm_points0, xmm_points1, xmm_result;
+
+  /*
+   * First do 4 values in every loop iteration.
+   * There may be up to 3 values left.
+   * leftovers0 indicates if at least 2 more are available for SSE execution.
+   * leftovers1 indicates if there is a single element left.
+   */
+  const int quarterPoints = num_points / 4;
+  const int leftovers0 = (num_points / 2) - 2 * quarterPoints;
+  const int leftovers1 = num_points % 2;
+
+  // load complex value into both parts of the register.
+  const __m128 xmm_symbol = _mm_castpd_ps(_mm_load1_pd((const double*)src0));
+  
+  // Load scalar into all 4 parts of the register
+  const __m128 xmm_scalar = _mm_load1_ps(&scalar);
+  
+  xmm_points0 = _mm_load_ps((float*)points);
+  xmm_points1 = _mm_load_ps((float*)(points + 2));
```

Comment:
These loads must not performed if ``num_points`` is smaller than 4 – out-of-bound memory access. Some day, on some machine this will fail with a SEGFAULT. Move into the loop below.